server/src/domain/library/library.service.ts

```diff
@@ -683,20 +680,45 @@ export class LibraryService extends EventEmitter {
       await this.assetRepository.updateAll(assetIdsToMarkOnline, { isOffline: false });
     }
 
-    if (!shouldScanAll) {
-      pathsToScan = [...crawledAssetPaths];
-      this.logger.debug(`Will import ${pathsToScan.length} new asset(s)`);
-    }
+    if (crawledAssetPaths.size > 0) {
+      if (!shouldScanAll) {
+        this.logger.debug(`Will import ${crawledAssetPaths.size} new asset(s)`);
+      }
+
+      const batch = [];
+      for (const assetPath of crawledAssetPaths) {
+        batch.push(assetPath);
 
-    if (pathsToScan.length > 0) {
-      await this.scanAssets(job.id, pathsToScan, library.ownerId, job.refreshAllFiles ?? false);
+        if (batch.length >= LIBRARY_SCAN_BATCH_SIZE) {
+          await this.scanAssets(job.id, batch, library.ownerId, job.refreshAllFiles ?? false);
+          batch.length = 0;
+        }
+      }
+
+      if (batch.length > 0) {
+        await this.scanAssets(job.id, batch, library.ownerId, job.refreshAllFiles ?? false);
+      }
     }
 
     await this.repository.update({ id: job.id, refreshedAt: new Date() });
 
     return true;
   }
 
+  private async getPathTrie(library: LibraryEntity, importPaths: string[]): Promise<Trie<string>> {
+    const generator = this.storageRepository.walk({
```

Comment:
It might make sense to move the library path validation in here too. Instead of passing it as a second property.

Reply:
Oh, good idea